File: PRNumber/app/src/main/java/com/pplus/prnumberuser/core/network/upload/AbstractUpload.java
package com.pplus.prnumberuser.core.network.upload;

import android.os.Handler;
import android.os.Looper;
import android.webkit.MimeTypeMap;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.pplus.utils.part.logs.LogUtil;
import com.pplus.utils.part.utils.StringUtils;
import com.pplus.prnumberuser.PRNumberApplication;
import com.pplus.prnumberuser.R;
import com.pplus.prnumberuser.apps.common.builder.AlertBuilder;
import com.pplus.prnumberuser.apps.common.builder.data.AlertData;
import com.pplus.prnumberuser.apps.common.mgmt.LoginInfoManager;
import com.pplus.prnumberuser.core.network.model.request.BaseParams;
import com.pplus.prnumberuser.core.network.model.request.params.ParamsAttachment;
import com.pplus.prnumberuser.core.network.model.response.NewResultResponse;
import com.pplus.prnumberuser.core.util.DebugConfig;
import com.pplus.prnumberuser.core.util.PplusCommonUtil;

import java.io.File;
import java.io.FileFilter;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

import okhttp3.Call;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.internal.Util;
import okio.Buffer;
import okio.BufferedSink;
import okio.Okio;
import okio.Source;

/**
 * Created by j2n on 2016. 9. 30..
 */

public abstract class AbstractUpload<T extends Object>{

    public final String LOG_TAG = getClass().getSimpleName();

    private OkHttpClient okHttpClient;

    private Map<String, Call> callMap;

    private LinkedBlockingQueue<Call> callQueue;

    private ThreadPoolExecutor threadPool;

    private PplusUploadListener<T> pplusCallback;

    /* 업로드 주소 */
    abstract String getUploadUrl();

    /* 멀티 쓰레드 사용여부*/
    abstract boolean isMultiThreadEnable();

    /**
     * new TypeToken<ResultResponse<T>>() {}.getType();
     */
    abstract TypeToken<NewResultResponse<T>> getResultType();

    protected AbstractUpload(PplusUploadListener<T> pplusCallback){

        //        this.okHttpClient

        OkHttpClient.Builder clientbuilder = new OkHttpClient.Builder();
        clientbuilder.connectTimeout(30, TimeUnit.SECONDS);
        clientbuilder.writeTimeout(180, TimeUnit.SECONDS);
        clientbuilder.readTimeout(30, TimeUnit.SECONDS);
        clientbuilder.addNetworkInterceptor(new Interceptor(){

            @Override
            public Response intercept(Chain chain) throws IOException{

                Request original = chain.request();

                Request.Builder builder = original.newBuilder();
                builder.method(original.method(), original.body());

                if(LoginInfoManager.getInstance().isMember()) {
                    LogUtil.e(LOG_TAG, "add session : {}", LoginInfoManager.getInstance().getUser().getSessionKey());
                    builder.addHeader("sessionKey", LoginInfoManager.getInstance().getUser().getSessionKey());
                }

                return chain.proceed(builder.build());
            }
        });

        clientbuilder.retryOnConnectionFailure(true);

        this.okHttpClient = clientbuilder.build();

        this.pplusCallback = pplusCallback;

        this.callMap = new HashMap<>();

        this.callQueue = new LinkedBlockingQueue<>();

        //        if(isMultiThreadEnable()) {
        //            this.maxThreadCount = isMultiThreadEnable() ? 4 : 1;
        //        }

        //        2, 4, 60, TimeUnit.SECONDS
        //        this.threadPool = new ThreadPoolExecutor(Math.max(maxThreadCount / 2, 2), maxThreadCount, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
        int maxThreadCount = 1;

        if(isMultiThreadEnable()) {
            maxThreadCount = getNumCores();
        }


        this.threadPool = new ThreadPoolExecutor(maxThreadCount, maxThreadCount, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
    }


    public void request(String tag, BaseParams baseParams){

        run(new UploadRunnable(tag, baseParams));
    }

    /**
     * 요청중인 콜에대해 취소합니다.
     */
    public void stop(PplusUploadCancelListener pplusUploadCancelListener){

        Iterator<String> callIterator = callMap.keySet().iterator();

        Set<String> stringSet = new HashSet<>();

        while (callIterator.hasNext()) {
            String tag = callIterator.next();

            if(callMap.containsKey(tag)) {
                callMap.get(tag).cancel();
                callMap.remove(tag);

                stringSet.add(tag);
            }
        }

        pplusUploadCancelListener.onCancel(stringSet);
    }

    private void run(Runnable runnable){

        threadPool.execute(runnable);
    }

    private class UploadRunnable implements Runnable{

        private String tag;
        private BaseParams baseParams;

        public UploadRunnable(String tag, BaseParams baseParams){

            this.tag = tag;
            this.baseParams = baseParams;
        }

        @Override
        public void run(){


            //            synchronized(callMap) {
            try {

                if(baseParams instanceof ParamsAttachment) {
                    synchronized(baseParams) {
                        ((ParamsAttachment) baseParams).buildFile();
                    }
                }

                Call call = okHttpClient.newCall(getRequest(baseParams.getParamsToMap()));

                callMap.put(tag, call);

                Response response = call.execute();

                if(DebugConfig.isDebugMode()) {

                    try {
                        LogUtil.e(LOG_TAG, "Successful = {} ", response.isSuccessful());

                        LogUtil.e(LOG_TAG, "request Url = {} ", response.request().url().toString());

                        LogUtil.e(LOG_TAG, "request header = {} ", response.request().header("sessionKey"));

                        final RequestBody copy = response.request().body();

                        if(copy != null) {
                            final Buffer buffer = new Buffer();
                            copy.writeTo(buffer);
                            LogUtil.e(LOG_TAG, "request body = {} ", buffer.readUtf8());
                        }

                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }

                if(!response.isSuccessful()) {

                    NewResultResponse resultResponse = new NewResultResponse();
                    resultResponse.setResultCode(500);

                    onFailure(tag, resultResponse);
                } else {

                    String responseResult = response.body().string();

                    GsonBuilder gsonBuilder = new GsonBuilder();
                    Gson gson = gsonBuilder.serializeNulls().disableHtmlEscaping().create();

                    Type collectionType = getResultType().getType();

                    NewResultResponse<T> resultResponse = gson.fromJson(responseResult, collectionType);

                    if(resultResponse.getResultCode() == 200) {
//                        int width = ((ParamsAttachment) baseParams).getWidth();
//                        int height = ((ParamsAttachment) baseParams).getHeight();
//                        ((Attachment) resultResponse.getData()).setWidth(width);
//                        ((Attachment) resultResponse.getData()).setHeight(height);
                        onResult(tag, resultResponse);
                    } else {
                        onFailure(tag, resultResponse);
                    }

                }

            } catch (IOException e) {
                e.printStackTrace();

                NewResultResponse resultResponse = new NewResultResponse();
                resultResponse.setResultCode(500);
                onFailure(tag, resultResponse);

            } finally {
                if(callMap.containsKey(tag)) callMap.remove(tag);

            }
            //            }
        }

        // 성공 처리
        void onResult(final String tag, final NewResultResponse<T> resultResponse){

            new Handler(Looper.getMainLooper()).post(new Runnable(){

                @Override
                public void run(){

                    pplusCallback.onResult(tag, resultResponse);
                }
            });
        }

        // 실패 처리
        void onFailure(final String tag, final NewResultResponse resultResponse){

            new Handler(Looper.getMainLooper()).post(new Runnable(){

                @Override
                public void run(){

                    pplusCallback.onFailure(tag, resultResponse);
                }
            });
        }
    }

    private Request getRequest(Map<String, Object> value){

        RequestBody requestBody = null;

        okhttp3.MultipartBody.Builder builder = new okhttp3.MultipartBody.Builder().setType(okhttp3.MultipartBody.FORM);

        Iterator<String> keys = value.keySet().iterator();

        while (keys.hasNext()) {
            String key = keys.next();

            Object object = value.get(key);

            if(object != null) {

                if(object instanceof File) {

                    final File file = (File) object;

                    //                    okhttp3.RequestBody fileBody = okhttp3.RequestBody.create(okhttp3.MediaType.parse(getMimeType(file)), file);

                    RequestBody fileBody = new RequestBody(){

                        @Override
                        public MediaType contentType(){

                            return MediaType.parse(getMimeType(file));
                        }

                        @Override
                        public long contentLength(){

                            return file.length();
                        }

                        @Override
                        public void writeTo(BufferedSink sink) throws IOException{

                            Source source = null;
                            try {
                                source = Okio.source(file);
                                sink.writeAll(source);
                                LogUtil.e(LOG_TAG, "file upload... source = {}", source);
                            } finally {
                                Util.closeQuietly(source);
                            }
                        }
                    };

                    builder.addFormDataPart(key, file.getName(), fileBody);

                } else if(object instanceof String) {
                    builder.addFormDataPart(key, (String) value.get(key));
                } else if(object instanceof Enum) {
                    builder.addFormDataPart(key, ((Enum) value.get(key)).name());
                } else if(object instanceof Long || object instanceof Integer || object instanceof Float) {
                    builder.addFormDataPart(key, String.valueOf(value.get(key)));
                }
            }
        }

        requestBody = builder.build();

        return new Request.Builder().url(getUploadUrl()).post(requestBody).build();
    }


    public String getMimeType(File file){

        return getMimeType(file.getPath());
    }

    public String getMimeType(String url){

        String type = null;

        String extension = MimeTypeMap.getFileExtensionFromUrl(url);
        if(extension != null) {
            type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
        }

        if(StringUtils.isEmpty(type)) {
            type = "*/*";
        }

        return type;
    }

    public void checkSelfNetwork(){

        if(!PplusCommonUtil.Companion.hasNetworkConnection()) {
            AlertBuilder.Builder builder = new AlertBuilder.Builder();
            builder.setTitle(PRNumberApplication.getContext().getString(R.string.word_notice_alert));
            builder.setStyle_alert(AlertBuilder.STYLE_ALERT.MESSAGE);
            builder.addContents(new AlertData.MessageData(PRNumberApplication.getContext().getString(R.string.msg_disconnected_network), AlertBuilder.MESSAGE_TYPE.TEXT, 1));
            builder.addContents(new AlertData.MessageData(PRNumberApplication.getContext().getString(R.string.msg_check_network_status), AlertBuilder.MESSAGE_TYPE.TEXT, 2));
            builder.setLeftText(PRNumberApplication.getContext().getString(R.string.word_confirm));
            builder.builder().show(PRNumberApplication.getContext(), true);
            return;
        }
    }

    private int getNumCores(){

        //Private Class to display only CPU devices in the directory listing

        class CpuFilter implements FileFilter{

            @Override

            public boolean accept(File pathname){

                //Check if filename is "cpu", followed by a single digit number

                if(Pattern.matches("cpu[0-9]", pathname.getName())) {

                    return true;

                }

                return false;

            }

        }


        try {

            //Get directory containing CPU info

            File dir = new File("/sys/devices/system/cpu/");

            //Filter to only list the devices we care about

            File[] files = dir.listFiles(new CpuFilter());

            //Return the number of cores (virtual CPU devices)

            return files.length;

        } catch (Exception e) {

            return 1;

        }

    }
}

File: WowBox/app/src/main/java/com/lejel/wowbox/core/network/upload/AbstractUpload.kt
package com.lejel.wowbox.core.network.upload

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.ExifInterface
import android.os.Handler
import android.os.Looper
import android.webkit.MimeTypeMap
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.pplus.utils.part.logs.LogUtil
import com.pplus.utils.part.utils.StringUtils
import com.lejel.wowbox.Const
import com.lejel.wowbox.apps.common.mgmt.LoginInfoManager
import com.lejel.wowbox.core.network.model.response.NewResultResponse
import com.lejel.wowbox.core.util.PplusCommonUtil
import com.lejel.wowbox.core.util.PplusCommonUtil.Companion.getDeviceID
import com.lejel.wowbox.core.util.imageresizer.ImageResizer
import com.lejel.wowbox.core.util.imageresizer.operations.ImageRotation
import okhttp3.Call
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.internal.closeQuietly
import okio.Buffer
import okio.BufferedSink
import okio.Source
import okio.source
import java.io.File
import java.io.FileFilter
import java.io.IOException
import java.lang.reflect.Type
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

/**
 * Created by j2n on 2016. 9. 30..
 */
abstract class AbstractUpload<T : Any?> protected constructor(pplusCallback: PplusUploadListener<T>) {
    val LOG_TAG = javaClass.simpleName
    private val okHttpClient: OkHttpClient
    private val callMap: MutableMap<String, Call>
    private val callQueue: LinkedBlockingQueue<Call>
    private val threadPool: ThreadPoolExecutor
    private val pplusCallback: PplusUploadListener<T>

    /* 업로드 주소 */
    abstract fun getUploadUrl(): String

    /* 멀티 쓰레드 사용여부*/
    abstract fun isMultiThreadEnable(): Boolean
    abstract fun getResultType(): TypeToken<NewResultResponse<T>>

    init {

        //        this.okHttpClient
        val clientbuilder = OkHttpClient.Builder()
        clientbuilder.connectTimeout(30, TimeUnit.SECONDS)
        clientbuilder.writeTimeout(180, TimeUnit.SECONDS)
        clientbuilder.readTimeout(30, TimeUnit.SECONDS)
        clientbuilder.addNetworkInterceptor(Interceptor { chain ->
            val original = chain.request()
            val builder = original.newBuilder()
            builder.method(original.method, original.body)
            if (LoginInfoManager.getInstance().isMember()) {
                LoginInfoManager.getInstance().member!!.token?.let { builder.addHeader("token", it) }
            }
            builder.addHeader("device", getDeviceID())
            builder.addHeader("api_key", "Rx4FTCJDi3MwyquI")
            chain.proceed(builder.build())
        })
        clientbuilder.retryOnConnectionFailure(true)
        okHttpClient = clientbuilder.build()
        this.pplusCallback = pplusCallback
        callMap = HashMap()
        callQueue = LinkedBlockingQueue()
        var maxThreadCount = 1
        if (isMultiThreadEnable()) {
            maxThreadCount = getNumCores()
        }
        threadPool = ThreadPoolExecutor(maxThreadCount, maxThreadCount, 60, TimeUnit.SECONDS, LinkedBlockingQueue())
    }

    fun request(tag: String, params: HashMap<String, Any>) {
        run(UploadRunnable(tag, params))
    }

    /**
     * 요청중인 콜에대해 취소합니다.
     */
    fun stop(pplusUploadCancelListener: PplusUploadCancelListener) {
        val callIterator: Iterator<String> = callMap.keys.iterator()
        val stringSet: MutableSet<String> = HashSet()
        while (callIterator.hasNext()) {
            val tag = callIterator.next()
            if (callMap.containsKey(tag)) {
                callMap[tag]!!.cancel()
                callMap.remove(tag)
                stringSet.add(tag)
            }
        }
        pplusUploadCancelListener.onCancel(stringSet)
    }

    private fun run(runnable: Runnable) {
        threadPool.execute(runnable)
    }

    private inner class UploadRunnable(private val tag: String, private val params: HashMap<String, Any>) : Runnable {
        override fun run() {


            //            synchronized(callMap) {
            try {
                synchronized(params) {
                    params["file"] = buildFile(params["file"] as File)
                }
                val call = okHttpClient.newCall(getRequest(params))
                callMap[tag] = call
                val response = call.execute()
                if (Const.DEBUG_MODE) {
                    try {
                        LogUtil.e(LOG_TAG, "Successful = {} ", response.isSuccessful)
                        LogUtil.e(LOG_TAG, "request Url = {} ", response.request.url.toString())
                        val copy = response.request.body
                        if (copy != null) {
                            val buffer = Buffer()
                            copy.writeTo(buffer)
//                            LogUtil.e(LOG_TAG, "request body = {} ", buffer.readUtf8())
                        }
                    } catch (e: IOException) {
                        LogUtil.e(LOG_TAG, e.toString())
                    }
                }
                if (!response.isSuccessful) {
                    val resultResponse: NewResultResponse<*> = NewResultResponse<Any?>()
                    resultResponse.code = 500
                    onFailure(tag, resultResponse)
                } else {
                    val responseResult = response.body!!.string()
                    val gsonBuilder = GsonBuilder()
                    val gson = gsonBuilder.serializeNulls().disableHtmlEscaping().create()
                    val collectionType: Type = getResultType().type
                    val resultResponse = gson.fromJson<NewResultResponse<T>>(responseResult, collectionType)
                    if (resultResponse.code == 200) {
                        onResult(tag, resultResponse)
                    } else {
                        onFailure(tag, resultResponse)
                    }
                }
            } catch (e: IOException) {
                LogUtil.e(LOG_TAG, e.toString())
                val resultResponse: NewResultResponse<*> = NewResultResponse<Any?>()
                resultResponse.code = 500
                onFailure(tag, resultResponse)
            } finally {
                if (callMap.containsKey(tag)) callMap.remove(tag)
            }
        }

        // 성공 처리
        fun onResult(tag: String?, resultResponse: NewResultResponse<T>?) {
            Handler(Looper.getMainLooper()).post { pplusCallback.onResult(tag, resultResponse) }
        }

        // 실패 처리
        fun onFailure(tag: String?, resultResponse: NewResultResponse<*>?) {
            Handler(Looper.getMainLooper()).post { pplusCallback.onFailure(tag, resultResponse) }
        }
    }

    fun buildFile(file: File): File {
        val path = file.path
        val originFile = File(path)
        var imageRotation: ImageRotation? = null
        try {
            val exif = ExifInterface(path)
            val orientation = exif.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_UNDEFINED)
            when (orientation) {
                ExifInterface.ORIENTATION_ROTATE_90 -> imageRotation = ImageRotation.CW_90

                ExifInterface.ORIENTATION_ROTATE_180 -> imageRotation = ImageRotation.CW_180

                ExifInterface.ORIENTATION_ROTATE_270 -> imageRotation = ImageRotation.CW_270
            }
        } catch (e: IOException) {
            LogUtil.e(LOG_TAG, e.toString())
        }
        val tempFile = File(PplusCommonUtil.getAlbumDir(), originFile.name)
        if (!tempFile.exists()) {
            try {
                tempFile.createNewFile()
            } catch (e: IOException) {
                LogUtil.e(LOG_TAG, e.toString())
            }
        }
        var bitmap: Bitmap? = null //        if (isResize) { //            try {
        //                bitmap = if (imageRotation != null) {
        //                    when (imageRotation) {
        //                        ImageRotation.CW_90, ImageRotation.CW_180 -> ImageResizer.resize(originFile, 1280, 720, ResizeMode.FIT_TO_HEIGHT)
        //                        else -> ImageResizer.resize(originFile, 720, 1280, ResizeMode.FIT_TO_WIDTH)
        //                    }
        //                } else {
        //                    ImageResizer.resize(originFile, 720, 1280, ResizeMode.FIT_TO_WIDTH)
        //                }
        //                if (bitmap == null) {
        //                    LogUtil.e(LOG_TAG, "bitmap is null")
        //                }
        //                ImageResizer.saveToFile(bitmap, tempFile)
        //            } catch (error: OutOfMemoryError) {
        //                LogUtil.e(LOG_TAG, "error = {}", error)
        //            } finally {
        //                bitmap?.recycle()
        //            }
        //        }
        if (imageRotation != null) {
            try {
                bitmap = ImageResizer.rotate(tempFile, imageRotation)
                ImageResizer.saveToFile(bitmap, tempFile)
            } catch (error: OutOfMemoryError) {
                LogUtil.e(LOG_TAG, "error = {}", error)
            } finally {
                bitmap?.recycle()
            }
        }
        val options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        BitmapFactory.decodeFile(tempFile.path, options)
        bitmap?.recycle()

        // 최종 파일 변경
        return tempFile
    }

    private fun getRequest(params: Map<String, Any>): Request {
        var requestBody: RequestBody? = null
        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        val keys = params.keys.iterator()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = params[key]
            if (value != null) {
                if (value is File) {

                    val fileBody: RequestBody = object : RequestBody() {
                        override fun contentType(): MediaType? {
                            return getMimeType(value)?.toMediaTypeOrNull()
                        }

                        override fun contentLength(): Long {
                            return value.length()
                        }

                        @Throws(IOException::class)
                        override fun writeTo(sink: BufferedSink) {
                            var source: Source? = null
                            try {
                                source = value.source()
                                sink.writeAll(source)
                                LogUtil.e(LOG_TAG, "file upload... source = {}", source)
                            } finally {
                                source!!.closeQuietly()
                            }
                        }
                    }
                    builder.addFormDataPart(key, value.name, fileBody)
                } else if (value is String) {
                    builder.addFormDataPart(key, params[key] as String)
                } else if (value is Enum<*>) {
                    builder.addFormDataPart(key, (params[key] as Enum<*>?)!!.name)
                } else if (value is Long || value is Int || value is Float) {
                    builder.addFormDataPart(key, params[key].toString())
                }
            }
        }
        requestBody = builder.build()
        return Request.Builder().url(getUploadUrl()).post(requestBody).build()
    }

    fun getMimeType(file: File): String? {
        return getMimeType(file.path)
    }

    fun getMimeType(url: String?): String? {
        var type: String? = null
        val extension = MimeTypeMap.getFileExtensionFromUrl(url)
        if (extension != null) {
            type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension)
        }
        if (StringUtils.isEmpty(type)) {
            type = "*/*"
        }
        return type
    }


    //Check if filename is "cpu", followed by a single digit number


    //Get directory containing CPU info

    //Filter to only list the devices we care about

    //Return the number of cores (virtual CPU devices)
    //Private Class to display only CPU devices in the directory listing
    private fun getNumCores(): Int {

        //Private Class to display only CPU devices in the directory listing
        class CpuFilter : FileFilter {
            override fun accept(pathname: File): Boolean {

                //Check if filename is "cpu", followed by a single digit number
                return Pattern.matches("cpu[0-9]", pathname.name)
            }
        } //Get directory containing CPU info

        val dir = File("/sys/devices/system/cpu/")

        //Filter to only list the devices we care about
        val files = dir.listFiles(CpuFilter())

        //Return the number of cores (virtual CPU devices)
        return files?.size ?: 1
    }
}